opencti-platform/opencti-graphql/src/database/engine.js

```diff
@@ -3586,24 +3603,109 @@ const prepareIndexingElement = async (thing) => {
   const entity = prepareEntity(thing);
   return prepareElementForIndexing(entity);
 };
-const prepareIndexing = async (elements) => {
+const prepareIndexing = async (context, user, elements) => {
+  const draftContext = getDraftContext(context, user);
   const preparedElements = [];
   for (let i = 0; i < elements.length; i += 1) {
     const element = elements[i];
+    if (draftContext) {
+      // If we are in a draft, relations from and to need to be elements that are also in draft.
+      if (element.base_type === BASE_TYPE_RELATION) {
+        const { from, to } = element;
+        if (!elements.some((e) => e.internal_id === from.internal_id)) {
+          const draftFrom = await loadDraftElement(context, user, from);
+          element.from = draftFrom;
+          element.fromId = draftFrom.id;
+        } else {
+          element.from._index = INDEX_DRAFT_OBJECTS;
+        }
+        if (!elements.some((e) => e.internal_id === to.internal_id)) {
+          const draftTo = await loadDraftElement(context, user, to);
+          element.to = draftTo;
+          element.toId = draftTo.id;
+        } else {
+          element.to._index = INDEX_DRAFT_OBJECTS;
+        }
+      }
+      element._index = INDEX_DRAFT_OBJECTS;
+      element.draft_ids = [draftContext];
+    }
     const prepared = await prepareIndexingElement(element);
     preparedElements.push(prepared);
   }
   return preparedElements;
 };
+export const elListExistingDraftWorkspaces = async (context, user) => {
+  const listArgs = {
+    connectionFormat: false,
+    filters: { mode: FilterMode.And, filters: [{ key: ['entity_type'], values: [ENTITY_TYPE_DRAFT_WORKSPACE] }], filterGroups: [] }
+  };
+  return elList(context, user, READ_INDEX_INTERNAL_OBJECTS, listArgs);
+};
+// Creates a copy of a live element in the draft index with the current draft context
+const copyLiveElementToDraft = async (context, user, element) => {
+  const draftContext = getDraftContext(context, user);
+  if (!draftContext || element._index.includes(INDEX_DRAFT_OBJECTS)) return element;
+
+  const updatedElement = structuredClone(element);
+  const newId = generateInternalId();
+  const reindexOpts = { dbId: newId, sourceUpdate: { draft_ids: [draftContext] } };
+  await elReindexElements(context, user, [element.internal_id], element._index, INDEX_DRAFT_OBJECTS, reindexOpts);
+  updatedElement._id = newId;
+  updatedElement._index = INDEX_DRAFT_OBJECTS;
+
+  // Add draftId to live element draftsIds
+  const allDrafts = await elListExistingDraftWorkspaces(context, user);
```

Comment:
I think we should list all drafts with SYSTEM user, to make sure we have access to all drafts (even though for now drafts are accessible to anyone, but technically we might have access restrictions later)